document-store/src/integrationTest/java/org/hypertrace/core/documentstore/DocStoreQueryV1Test.java

```diff
@@ -1977,8 +1982,201 @@ public void testAtomicUpdateDocumentWithoutSelections(final String datastoreName
+          iterator_new,
+          "query/update_operator/updated_removes_all_occurrances_from_integer_list.json",
+          9);
+    }
+
+    @ParameterizedTest
+    @ArgumentsSource(AllProvider.class)
+    void testAddToListIfAbsentDoesNotDeduplicateTheExistingList(final String datastoreName)
+        throws IOException {
+      final Collection collection = getCollection(datastoreName, UPDATABLE_COLLECTION_NAME);
+      createCollectionData("query/updatable_collection_data.json", UPDATABLE_COLLECTION_NAME);
+
+      final SubDocumentUpdate unset =
+          SubDocumentUpdate.builder().subDocument("sales").operator(UNSET).build();
+      final SubDocumentUpdate add =
+          SubDocumentUpdate.builder()
+              .subDocument("props.added.list")
+              .subDocumentValue(SubDocumentValue.of(new Integer[] {5, 1, 5}))
+              .build();
+
+      final Query query = Query.builder().build();
+      final List<SubDocumentUpdate> updates = List.of(add, unset);
+
+      final CloseableIterator<Document> iterator =
+          collection.bulkUpdate(
+              query, updates, UpdateOptions.builder().returnDocumentType(NONE).build());
+
+      final SubDocumentUpdate remove =
+          SubDocumentUpdate.builder()
+              .subDocument("props.added.list")
+              .operator(ADD_TO_LIST_IF_ABSENT)
+              .subDocumentValue(SubDocumentValue.of(new Integer[] {3, 1, 4}))
+              .build();
+      final List<SubDocumentUpdate> new_updates = List.of(remove);
+
+      final CloseableIterator<Document> iterator_new =
+          collection.bulkUpdate(
+              query, new_updates, UpdateOptions.builder().returnDocumentType(AFTER_UPDATE).build());
+      assertDocsAndSizeEqualWithoutOrder(
+          datastoreName,
+          iterator_new,
+          "query/update_operator/updated_add_to_list_if_absent_does_not_deduplicate_existing_list.json",
+          9);
+    }
+
+    @ParameterizedTest
+    @ArgumentsSource(AllProvider.class)
+    void testSameHierarchyUpdateThrowsException(final String datastoreName) throws IOException {
+      final Collection collection = getCollection(datastoreName, UPDATABLE_COLLECTION_NAME);
+      createCollectionData("query/updatable_collection_data.json", UPDATABLE_COLLECTION_NAME);
+
+      final SubDocumentUpdate unset =
+          SubDocumentUpdate.builder().subDocument("props.added").operator(UNSET).build();
+      final SubDocumentUpdate add =
+          SubDocumentUpdate.builder()
+              .subDocument("props.added.list")
+              .subDocumentValue(SubDocumentValue.of(new Integer[] {5, 1, 5}))
+              .build();
+
+      final Query query = Query.builder().build();
+      final List<SubDocumentUpdate> updates = List.of(add, unset);
+
+      assertThrows(
+          IOException.class,
+          () ->
+              collection.bulkUpdate(
+                  query, updates, UpdateOptions.builder().returnDocumentType(NONE).build()));
```

Comment:
Why should it throw an exception?
- First, we are creating the property - `props.added.list`
- Second, its removes `props.added`
So, what is not allowed - a partial path?

Reply:
If we allow such paths, we need to ensure that MongoDB and Postgres run those updates in the same order.
Basically, the result would become arbitrary breaking the compatibility between MongoDB and Postgres.

In other words, this is to let the clients know that the updates can be executed in any order, which is why it takes in a `Collection` rather than a list.

Reply:
To elaborate a bit on your example, 
If SET `props.added.list` gets executed first, and then UNSET `props.added`, the result would be without the `props.added` path.
However, if we reverse the order, then `props.added.list` would be set. 
Basically, it creates confusion depending on the execution order.

If they clients really want such an update, they will have to invoke the `update` API twice once for each path one after the other so that they have clarity on the execution order.